epsagon/trace.py

```diff
@@ -307,43 +306,49 @@ def _pop_trace_async_mode(self):
 
     def _get_trace(self, unique_id=None, should_create=False):
         """
-        Get trace based on the use_single_trace/use_async_tracer.
+        Get trace based on the tracing mode:
+        single trace (use_single_trace=True) or
+        multiple threads (use_single_trace=False) or
+        async tasks (use_async_tracer=True)
+
         If should_create then creating a new trace if trace does not exist
         if use_single_trace is set to False, each thread will have
-        it's own trace.
+        it's own trace..
         if use_async_tracer is set to True, each asyncio.Task will be
         assigned with the trace.
         :return: The trace.
         """
-        if self.use_async_tracer:
-            return self._get_tracer_async_mode(should_create=should_create)
-
-        unique_id = self.get_thread_local_unique_id(unique_id)
-        if unique_id:
-            trace = (
-                self.singleton_trace
-                if self.singleton_trace and not self.traces
-                else self.traces.get(
-                    unique_id, None
+        with TraceFactory.LOCK:
```

Comment:
Can you double check to make sure that either this is a reentrant lock or that there are no internal calls that also use that lock?